### DISCUSSION TITLE
Azure functions and correlation

### DISCUSSION THREAD

**Discussion by DeAbdy** [arcus-azure/arcus.observability#519] [General]
Hi ,

I am new to the Arcus platform and I have a question about correlation. 
I have a few azure functions (http, blobtrigger, queue) that connects in order to each other but all works async. In this flow I want to correlate the transaction.

How can do that correlation? Is their a custom way? Or is this not yet build?

**Comment by stijnmoreels** [comment]
Hi @DeAbdy ! Amazing to hear that your are investigating in the Arcus platform. Great to have you!

We have recently released a release across all the Arcus components that now works seamlessly together with Microsoft telemetry.
Here are some links to Azure Functions-specific pages to learn more about how telemetry is set up in an Arcus project. Also note that we have several [Arcus project templates](https://templates.arcus-azure.net/) which soon will receive this new update.

* [Azure Functions in-process/isolated HTTP trigger](https://webapi.arcus-azure.net/Features/correlation-azure-functions)
* [Azure Functions Service Bus trigger](https://messaging.arcus-azure.net/Features/message-handling/service-bus-azure-functions)
* [Service-to-service correlation](https://observability.arcus-azure.net/Guidance/Service-to-service%20Correlation/index)

Don't hesitate to reach out if anything is unclear!

**Reply by DeAbdy** [reply]
Hi Stijn,

Thanks I have read this documentation but it is still unclear for me how I can correlate messages into a functional transaction.

example of one of the flows.

On a azurefunction with blobtrigger comes an message this i want to register as the start. Then this message is putting on a queue. But then the correlation is not passed because it goes into a new process. I want to know how can I correlate these message with custom calls. I can pass the requested/start correlationId?

**Reply by stijnmoreels** [reply]
If you register the regular Microsoft Application Insights, and use the Azure clients in your Azure Functions blob trigger, that would be enough. Note that at this point, there's no Arcus involved as Microsoft provides default correlation for this.
When the Application Insights services are registered in your application, it will enrich your Service Bus queue message with a `Diagnostic-Id` which contains this correlation information.

**Reply by stijnmoreels** [reply]
Adding @pim-simons to the mix as he has a lot more experience with Azure Functions.

**Reply by pim-simons** [reply]
I have some examples of scheduled functions where I use correlation to link everything together.
Firstly I define a transactionId like this:
``` c#
string transactionId = Guid.NewGuid().ToString();
```

I then use this transactionId value in all my logging by adding it to the telemetryContext:
``` c#
var telemetryContext = new Dictionary<string, object>
{
	{ "TransactionId", transactionId },
	{ "OperationId", Guid.NewGuid().ToString() }
};
```

And then adding the telemetryContext to my logging statements:
``` c#
Logger.LogSqlDependency(sqlConnectionStringBuilder.DataSource, "DummyDatabase", "DummyCommand", "DummyOperation", isSuccessful: isSqlSuccess, startTime: sqlStartTime, duration: sqlDurationMeasurement.Elapsed, Guid.NewGuid().ToString(), telemetryContext);
``` 

Does this answer your question?

**Reply by DeAbdy** [reply]
@pim-simons Probably yes. So then I pass the transactionId to my next operation and create a new operationId there and log the dependency.

I will try this. This will take me some time to change.

**Reply by DeAbdy** [reply]
@pim-simons @stijnmoreels I have implemented one flow of my application with the transactionid. I can now notice a end-to-end transaction. I only did some happy flow so other error logging still needs to be done.

I have only one question left. Is there a way to connect these dependencies to a current Request where it all starts? Or do I then need to log a customrequest with the same context?

Thanks.

**Reply by pim-simons** [reply]
Good to hear you have the end-to-end working in the happy flow 👍🏻 
I usually add this statement in the `finally` part of my overall try-catch:
``` c#
Logger.LogRequest(request, (int)httpStatusCode, durationMeasurement, telemetryContext);
```
And then make sure that the `telemetryContext` contains the TransactionId. But this is an example for an HTTP Triggered Azure Function. I don't have any experience personally with the blob trigger, @stijnmoreels  maybe you have some experience there?

**Reply by stijnmoreels** [reply]
I would think actually that adding this would do the trick as Application Insights is normally added by default by Microsoft, as is request tracking with W3C. Only for Azure Functions blob triggers would it be necessary to add this, I guess, as it is not by default something we need to support.

```csharp
services.AddLogging(logging => logging.AddApplicationInsightsWebJobs());
```

But, for Azure Functions HTTP triggers we do have our own middleware component for you to add, and for Azure Service bus triggers we also have this added internally.

**Reply by DeAbdy** [reply]
@stijnmoreels & @pim-simons Thanks. 

I tried both and they will all help with the outcome. But to make it more easier to quickly search up a functional transaction I have created a custom event. In the context of all my calls I put a transactionId. In application Insights this will make sure everything gets related. Like Requests or dependencies. I did not add the lower level actions because then it will becomes to fuzzy in my opinion.

**Reply by stijnmoreels** [reply]
Great that you got something to work with! It's super to have a correlation model of your application for numerous reasons, and if Arcus can help a little, it would be our pleasure. 😄
